### PR TITLE
hotreload: add component callback support

### PR DESCRIPTION
**Pull request by lrascao**
# Description

Component callback is a useful dapr integration functionality for applications to react on when components are loaded. This patch bridges the gap left between this and hot reload providing the same experience on both scenarios.

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [X] Code compiles correctly
- [ ] Created/updated tests
- [ ] Unit tests passing
- [ ] End-to-end tests passing
- [ ] Extended the documentation / Created issue in the <https://github.com/dapr/docs/> repo: dapr/docs#_[issue number]_
- [ ] Specification has been updated / Created issue in the <https://github.com/dapr/docs/> repo: dapr/docs#_[issue number]_
- [ ] Provided sample for the feature / Created issue in the <https://github.com/dapr/docs/> repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/runtime/hotreload/reconciler/components.go [hunk 1]
```diff
@@ -68,6 +70,14 @@ func (c *components) update(ctx context.Context, comp compapi.Component) {
 	if c.proc.AddPendingComponent(ctx, comp) {
 		log.Infof("Component updated: %s", comp.LogName())
 		c.proc.WaitForEmptyComponentQueue()
+
+		if c.cb != nil {
+			if err := c.cb(registry.ComponentRegistry{
```
Comment (JoshVanL) [inline pkg/runtime/hotreload/reconciler/components.go:75]: The callbacks here are at the wrong level- they need to be inside the component processor in `Init` and `Close`. Please return errors.

https://github.com/dapr/dapr/blob/334ae9eea43d487a7b29a0e4aef904e3eba57a10/pkg/runtime/processor/components.go#L30
https://github.com/dapr/dapr/blob/334ae9eea43d487a7b29a0e4aef904e3eba57a10/pkg/runtime/processor/components.go#L51

**Review by JoshVanL — CHANGES_REQUESTED** [review]

File: pkg/runtime/hotreload/reconciler/components.go (hunk 1 above)
Comment (lrascao) [reply pkg/runtime/hotreload/reconciler/components.go]: updated

File: pkg/runtime/runtime.go [hunk 2]
```diff
@@ -215,6 +215,7 @@ func newDaprRuntime(ctx context.Context,
 		ActorsEnabled:   len(runtimeConfig.actorsService) > 0,
 		Registry:        runtimeConfig.registry,
 		ComponentStore:  compStore,
+		ComponentCb:     runtimeConfig.registry.ComponentsCallback(),
```
Comment (famarting) [inline pkg/runtime/runtime.go:218]: is it true that at some point in the runtime initialization components callback is set to nil? I think I recall something like that

**Comment by famarting** [comment]
love this btw

File: pkg/runtime/runtime.go (hunk 2 above)
Comment (JoshVanL) [reply pkg/runtime/runtime.go:218]: It’s a pain to have a nil checker for this. If we can, let’s make sure this is always at least a noop func

File: pkg/runtime/processor/components.go [hunk 3]
```diff
@@ -44,7 +45,19 @@ func (p *Processor) Init(ctx context.Context, comp componentsapi.Component) erro
 		return errors.Join(err, p.compStore.DropPendingComponent())
 	}
 
-	return p.compStore.CommitPendingComponent()
+	if err := p.compStore.CommitPendingComponent(); err != nil {
+		return fmt.Errorf("error committing component: %w", err)
+	}
+
+	if p.componentCb != nil {
+		if err := p.componentCb(registry.ComponentRegistry{
```
Comment (JoshVanL) [inline pkg/runtime/processor/components.go:53]: This is not capturing init errors. I’m not sure if that’s intended.. we likely need to update the signature and wrap this init and close to pass that on.

**Review by JoshVanL — CHANGES_REQUESTED** [review]

File: pkg/runtime/runtime.go (hunk 2 above)
Comment (famarting) [reply pkg/runtime/runtime.go:218]: ah no, ignore me, its `a.appHealthReady` what its set to nil

File: pkg/runtime/processor/components.go (hunk 3 above)
Comment (JoshVanL) [reply pkg/runtime/processor/components.go:53]: It also follows that we should close the component of callback fails.

**Comment by lrascao** [comment]
closing in favour of https://github.com/dapr/dapr/pull/8268

**Pull request closed without merging**